src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs

```diff
@@ -123,12 +122,25 @@ protected override Payload ConstructHeadersPayload(GetHeadersPayload getHeadersP
                 return null;
             }
 
-            var headers = new ProvenHeadersPayload();
-            foreach (ChainedHeader header in this.chain.EnumerateToTip(fork).Skip(1))
+            var provenHeadersPayload = new ProvenHeadersPayload();
+
+            // Do not return more than 2000 headers from the fork point.
+            ChainedHeader chainedHeaderToStartFrom = this.consensusManager.Tip;
+            if ((this.consensusManager.Tip.Height - fork.Height) > MaxItemsPerHeadersMessage)
+            {
+                // e.g. If fork = 3000 and tip is 6000 we need to start from block 5000.
+                var startFromHeight = (fork.Height + 1) + MaxItemsPerHeadersMessage;
+                chainedHeaderToStartFrom = this.consensusManager.Tip.GetAncestor(startFromHeight);
```

Comment:
Same here use `chainedHeaderToStartFrom` Tip might change as this is  not inside the lock of CT.

Reply:
Ah ok yes, fixed!